src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,160 @@
+    let eviction = Event<'Key * EvictionReason>()
+
+    [<CLIEvent>] member val CacheHit = cacheHit.Publish
+    [<CLIEvent>] member val CacheMiss = cacheMiss.Publish
+    [<CLIEvent>] member val Eviction = eviction.Publish
+
+    // Increase expected capacity by the percentage to evict, since we want to not resize the dictionary.
+    member val Store = ConcurrentDictionary<'Key, CachedEntity<'Key,'Value>>(options.LevelOfConcurrency, capacity)
+    member val ConditionalWeakTable = new ConditionalWeakTable<_, Weak<_>>();
+
+    member _.GetStats() =
+        {|
+            Capacity = options.MaximumCapacity
+            PercentageToEvict = options.PercentageToEvict
+            Strategy = options.Strategy
+            LevelOfConcurrency = options.LevelOfConcurrency
+            Count = this.Store.Count
+            LeastRecentlyAccesssed = this.Store.Values |> Seq.minBy _.LastAccessed |> _.LastAccessed
+            MostRecentlyAccesssed = this.Store.Values |> Seq.maxBy _.LastAccessed |> _.LastAccessed
+            LeastFrequentlyAccessed = this.Store.Values |> Seq.minBy _.AccessCount |> _.AccessCount
+            MostFrequentlyAccessed = this.Store.Values |> Seq.maxBy _.AccessCount |> _.AccessCount
+        |}
+
+
+    member private _.GetEvictCount() =
+        if this.Store.Count >= options.MaximumCapacity then
+            this.Store.Count * options.PercentageToEvict / 100
+        else
+            0
+
+    // TODO: All of these are proofs of concept, a very naive implementation of eviction strategies, it will always walk the dictionary to find the items to evict, this is not efficient.
+    member private this.TryGetItemsToEvict () =
+        match options.Strategy with
+        | CachingStrategy.LRU ->
+            this.Store.Values |> Seq.sortBy _.LastAccessed |> Seq.take (this.GetEvictCount()) |> Seq.map (fun x -> x.Key)
+        | CachingStrategy.MRU ->
+            this.Store.Values |> Seq.sortByDescending _.LastAccessed |> Seq.take (this.GetEvictCount()) |> Seq.map (fun x -> x.Key)
+        | CachingStrategy.LFU ->
+            this.Store.Values |> Seq.sortBy _.AccessCount |> Seq.take (this.GetEvictCount()) |> Seq.map (fun x -> x.Key)
+
+    member private this.TryEvictItems () =
+        if this.GetEvictCount() > 0 then
+            for key in this.TryGetItemsToEvict () do
+                let (removed, _) = this.Store.TryRemove(key)
+                if removed then
+                    eviction.Trigger(key, EvictionReason.Evicted)
+
+    member private this.TryEvictTask () =
+        // This will spin in the background trying to evict items.
+        // One of the issues is that if the delay is high (>100ms), it will not be able to evict items in time, and the cache will grow beyond the maximum capacity.
+        backgroundTask {
+            while not cts.Token.IsCancellationRequested do
+                if this.GetEvictCount() > 0 then
+                    this.TryEvictItems ()
+                // do! Task.Delay(100, cts.Token)
```

Comment:
Exponential backoff based on number of evicted items?

Did I evict 0 - can afford a longer delay.
A lot of stuff evicted - try again shortly after.

Reply:
Yes, that was my idea as well

Reply:
So, I made it based on the cache utilization. It will fluctuate from 0 to 1 seconds. Worst case scenario would be if cache is populated in bursts (and has max size way off), then it will resize. Wondering if it's good enough for now? @0101 @T-Gro thoughts?